### PR TITLE
feat: initial work to make v1 API compatible with SIP-40 and SIP-41

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

This PR is a first step in making the v1 API compatible with [SIP-40](https://github.com/apache/superset/issues/9194) and [SIP-41](https://github.com/apache/superset/issues/9298).

Currently, the API returns exceptions using helper methods from Flask-AppBuilder, eg:

```python
        try:
            TestConnectionDatabaseCommand(g.user, item).run()
            return self.response(200, message="OK")
        except DatabaseTestConnectionFailedError as ex:
            return self.response_422(message=str(ex))
        except SupersetErrorException as ex:
            return self.response(ex.status, message=ex.error.message)
```

These responses are simple JSON payloads, not compatible with the standard error format introduced in SIP-40. For example, when adding a database with a non-existent driver (`foo://`) this is returned:

```json
{"message": "Could not load database driver: BaseEngineSpec"}
```

And this is the message displayed in the frontend (in a toast):

> ERROR: Could not load database driver: BaseEngineSpec

To conform with SIP-41 I added <del>a new decorator called `handle_exception`</del> <ins>an error handler</ins> that captures exceptions and formats them according to SIP-40. This allows us to simplify the API logic, removing the need to capture exceptions when running commands (see the [new `test_connection` method](https://github.com/apache/superset/pull/13960/files#diff-f9deb9b63f5f2db38b811ff092cf90b0d4d558223f743bbbb5b203c203d2d590L607-L613) in this PR).

With the decorator, the backend now returns:

```json
{
	"errors": [
		{
			"message": "Could not load database driver: BaseEngineSpec",
			"error_type": "GENERIC_COMMAND_ERROR",
			"level": "error",
			"extra": {
				"issue_codes": [
					{
						"code": 1010,
						"message": "Issue 1010 - Superset found an error while running a command."
					}
				]
			}
		}
	]
}
```

And the message is displayed correctly in the frontend as before, in a toast, without requiring any modifications:

> ERROR: Could not load database driver: BaseEngineSpec

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

See above.

### TEST PLAN

Will add unit tests, I wanted to discuss this first.

### ADDITIONAL INFORMATION

- [X] Has associated issue: https://github.com/apache/superset/issues/11882
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by etr2460 — COMMENTED** [review]
So in this case, you've got the `DatabaseTestConnectionFailedError` that `test_connection` can raise. Should this be something that extends a `SupersetErrorException` or should it be caught in the `handle_exception` wrapper and then return a response with a constructed SupersetError of type `TEST_CONNECTION_PORT_CLOSED_ERROR` or `TEST_CONNECTION_HOST_DOWN_ERROR`?

**Comment by betodealmeida** [comment]
> So in this case, you've got the `DatabaseTestConnectionFailedError` that `test_connection` can raise. Should this be something that extends a `SupersetErrorException` or should it be caught in the `handle_exception` wrapper and then return a response with a constructed SupersetError of type `TEST_CONNECTION_PORT_CLOSED_ERROR` or `TEST_CONNECTION_HOST_DOWN_ERROR`?

Right, I actually have another PR where I started doing that:

```python
class DatabaseTestConnectionFailedError(SupersetErrorsException):
    message = _("Connection failed, please check your connection settings")
```

This way it will expose the proper errors (and additional errors that I'm adding, eg, `TEST_CONNECTION_INVALID_USERNAME_ERROR`).

File: superset/views/base_api.py [hunk 1]
```diff
@@ -84,6 +88,72 @@ def wraps(self: "BaseSupersetModelRestApi", *args: Any, **kwargs: Any) -> Respon
...
+            logger.warning(ex)
+            return json_errors_response(errors=[ex.error], status=ex.status)
+        except SupersetErrorsException as ex:
+            logger.warning(ex)
+            return json_errors_response(errors=ex.errors, status=ex.status)
+        except CommandInvalidError as ex:
+            logger.warning(ex)
+            return json_errors_response(
+                errors=[
+                    SupersetError(
```
Comment (dpgaspar) [inline superset/views/base_api.py:121]: Are we always going to throw HTTP 500 on these? would it make sense that the exception itself could identify the HTTP status code?

File: superset/databases/api.py [hunk 2]
```diff
@@ -604,13 +609,8 @@ def test_connection(  # pylint: disable=too-many-return-statements
...
             return self.response_400(message=error.messages)
-        try:
-            TestConnectionDatabaseCommand(g.user, item).run()
-            return self.response(200, message="OK")
-        except DatabaseTestConnectionFailedError as ex:
-            return self.response_422(message=str(ex))
-        except SupersetErrorException as ex:
-            return self.response(ex.status, message=ex.error.message)
+        TestConnectionDatabaseCommand(g.user, item).run()
+        return self.response(200, message="OK")
```
Comment (dpgaspar) [inline superset/databases/api.py:613]: the OpenAPI error responses are not correct now, on the `BaseSupersetModelRestApi` we can declare the new SIP-40 schema https://github.com/dpgaspar/Flask-AppBuilder/blob/master/flask_appbuilder/api/__init__.py#L296

**Comment by dpgaspar** [comment]
Sometime ago @john-bodley proposed to use this `Flask` capability: https://flask.palletsprojects.com/en/1.1.x/errorhandling/#registering. I like the idea, and it seems doable, that way we would not need to implement our own decorator for error handling

**Review by dpgaspar — COMMENTED** [review]
Sometime ago @john-bodley proposed to use this Flask capability: https://flask.palletsprojects.com/en/1.1.x/errorhandling/#registering. I like the idea, and it seems doable, that way we would not need to implement our own decorator for error handling

File: superset/databases/api.py (hunk 2 above)
Comment (betodealmeida) [reply superset/databases/api.py:613]: Good catch, I'll fix!

File: superset/views/base_api.py (hunk 1 above)
Comment (betodealmeida) [reply superset/views/base_api.py:121]: I don't think that command exceptions should have HTTP status codes, they should care only about business logic and be decoupled from the API.

Ideally the API layer should do the mapping between command exceptions to `SupersetError(s)Exception`. The reason why I added `CommandInvalidError` and `CommandException` is so we can gradually implement that mapping, while in the meantime any non-covered command exceptions return a 500.

**Comment by betodealmeida** [comment]
> Sometime ago @john-bodley proposed to use this Flask capability: https://flask.palletsprojects.com/en/1.1.x/errorhandling/#registering. I like the idea, and it seems doable, that way we would not need to implement our own decorator for error handling

I agree, I love that pattern. Let me give it a try, it might be easier since we register the error handler and then we can gradually change the APIs to raise `SupersetErrorException` and `SupersetErrorsException` (or `ComandException` initially, but I'd rather have all APIs only returning one of the first two).

File: docs/src/pages/docs/Miscellaneous/issue_codes.mdx [hunk 3]
```diff
@@ -113,3 +113,21 @@ The host might be down, and cannot be reached on the provided port.
...
+Superset found an error while running a command.
+```
+
+Something unexpected happened, and Superset encountered an error while
+running a command. Please reach out to your administrator.
+
+## Issue 1011
+
+```
+Superset found an unexpected error.
```
Comment (etr2460) [inline docs/src/pages/docs/Miscellaneous/issue_codes.mdx:129]: nit: i prefer `encountered` here to `found`

```suggestion
Superset encountered an error while running a command.
```

Something unexpected happened, and Superset encountered an error while
running a command. Please reach out to your administrator.

## Issue 1011

```
Superset encountered an unexpected error.
```

File: superset/errors.py [hunk 4]
```diff
@@ -135,6 +139,20 @@ class SupersetErrorType(str, Enum):
...
+            "code": 1010,
+            "message": _(
+                "Issue 1010 - Superset found an error while running a command."
+            ),
+        },
+    ],
+    SupersetErrorType.GENERIC_BACKEND_ERROR: [
+        {
+            "code": 1011,
+            "message": _("Issue 1011 - Superset found an unexpected error."),
```
Comment (etr2460) [inline superset/errors.py:153]: same comment here

```suggestion
                "Issue 1010 - Superset encountered an error while running a command."
            ),
        },
    ],
    SupersetErrorType.GENERIC_BACKEND_ERROR: [
        {
            "code": 1011,
            "message": _("Issue 1011 - Superset encountered an unexpected error."),
```

File: tests/databases/api_tests.py [hunk 5]
```diff
@@ -820,7 +820,21 @@ def test_test_connection_failed(self):
         self.assertEqual(rv.headers["Content-Type"], "application/json; charset=utf-8")
         response = json.loads(rv.data.decode("utf-8"))
         expected_response = {
-            "message": "Could not load database driver: BaseEngineSpec",
+            "errors": [
+                {
+                    "message": "Could not load database driver: BaseEngineSpec",
```
Comment (etr2460) [inline tests/databases/api_tests.py:825]: i guess ideally in the future these messages will become actual errors with details on how to resolve them in the documentation

**Review by etr2460 — APPROVED** [review]
this all makes sense to me, stamping from the frontend perspective

File: tests/databases/api_tests.py (hunk 5 above)
Comment (betodealmeida) [reply tests/databases/api_tests.py:825]: Right, we're going to work on all the messaging in the test connection flow and add database flow.

**Pull request merged**